text/00110-compatibility-strategy.md

```diff
@@ -0,0 +1,364 @@
+
+- https://github.com/aws/aws-cdk/issues/5986
+- https://github.com/aws/aws-cdk/issues/4897
+- https://github.com/aws/aws-cdk/issues/4294
+
+These issues were very awkward and caused both our customers and ourselves a lot of unwanted work.
+They also reflect poorly on the stability of the CDK, especially in a post GA era.
+
+*We want to implement this RFC to prevent (or at least greatly reduce) the occurrences of such issues.*
+
+## (2) Refactor CX Protocol versioning
+
+We've talked about the mechanism we currently use to perform protocol version changes quite a bit. The process and the expected output aren't very clear, and we seem to be always trying to understand it again and again.
+This might warrant a re-think/re-design, to make it much more ergonomic and developer friendly. 
+
+However, performing changes on such a crucial part of the system, shouldn't be done without a proper validation mechanism.
+
+*We want to implement this RFC so that we are able to iterate and change the versioning mechanism itself* 
+
+Actually, We have already acknowledged the need for this a while back:
+
+- https://github.com/aws/cdk-ops/issues/365
+- https://github.com/aws/cdk-ops/issues/233
+
+Its time we solve this once and for all.
+
+After implementing this RFC, we can expect the following outcomes:
+
+- An incompatible change in the CLI won't make it to the release.
+- An incompatible change in the framework, will always be accompanied with a version bump.
+- An incompatible change in the framework, accompanied with a protocol version bump, will always produce the correct error message.
+
+# Design Summary
+
+In order to ensure compatibility between the CLI and the Framework, we have to actually run the CLI against an old framework version, and vice versa.
+We want to find a generic mechanism for running these tests, without having to cherry-pick use-cases or versions. 
+
+The main design idea is to introduce new types of integration tests into our pre-release suite. Those tests will run various permutations of `(CLI X Framework)`, and validate the expected result in each permutation.
+
+To make it generic, the tests will run the entire current integration tests suite, and simply inject different versions each time.
+
+An important note is that when testing compatibility to older versions (be it CLI or Framework), we cannot run the *latest* integration test suite, as it may contain tests for new features and bug fixes, which were not implement in older versions.
+
+Therefore, the entire design relies on us having the ability to:
+
+- *Run **old** integration tests using a **new** CLI version and an **old** Framework version.*
+- *Run **old** integration tests using a **new** Framework version and an **old** CLI version.*
+- *Mark a Framework version as incompatible*
+
+    - The tests should support an ability to specify versions who are not compatible, since the expected output would be different.
```

Comment:
I actually would rather _not_ having to do this. If the test needs changing each time the code it tests changes... then it's a poor test, isn't it?

Reply:
Well, that depends. 

I like that notion as a general rule of thumb, as it is usually a code smell for the test not having the proper assertions. 

However, sometimes this requires a deeper look. Sometimes, simply modifying (touching) the test file, doesn't mean the test itself changes.

The necessary changes to the test file will not change the behavior and the expected overall result of the code. This is why it was very important to extract `EXPECTED_ERROR` to a generic, dynamic value that wouldn't be changed by developers when pushing new code. 

If we go with this mechanism, we need a way of either marking (manually) or detecting (automatically) these incompatibilities. You can have a look at the [Auto Detecting Incompatibilities](https://github.com/aws/aws-cdk-rfcs/blob/epolon/compatibility-strategy/text/00110-cli-framework-compatibility-strategy.md#auto-detecting-incompatibilities) I included as a future possibility. But to me this isn't mandatory because it has the same affect, and all it does is alleviate some manual work. 

I would agree that automating this has the benefit of reducing risk when editing the file, but i think we can make the editing phase super simple (maybe create a dedicated file for incompatible versions) and live peacefully with it. 

Having said that, if you feel that editing this file in this way is indicative of some gap in the assertion logic itself, that is definitely worth addressing.

Reply:
This kind of annual work can be automation-hostile, and come in the way of future automation effort. This is architectural or technical debt we're taking at this stage and it's important to carefully consider it before moving forward with it.

Any manual requirement is also one more thing that someone can forget to do, and that is going to possibly be difficult to navigate by new contributors.

I still stand on my assertion that, if you need to touch the test each time you touch the tested code, then this is a bad test (self-fulfilling prophecy / mere validation that the implementation is exactly what we wrote it to be).

Reply:
Not relevant anymore